src/dotty/tools/dotc/transform/Space.scala

```diff
@@ -54,221 +60,219 @@ trait SpaceLogic {
-      if (!isEqualType(t1, t2)) Empty
+    case (Kon(tp1, ss1), Kon(tp2, ss2)) =>
+      if (!isEqualType(tp1, tp2)) Empty
       else if (ss1.zip(ss2).exists(p => simplify(intersect(p._1, p._2)) == Empty)) Empty
       else
-        Kon(t1, ss1.zip(ss2).map((intersect _).tupled))
+        Kon(tp1, ss1.zip(ss2).map((intersect _).tupled))
   }
 
-  //  the space of a not covered by b
+  /** The space of a not covered by b */
   def minus(a: Space, b: Space): Space = (a, b) match {
     case (Empty, _) => Empty
     case (_, Empty) => a
-    case (Typ(t1), Typ(t2)) =>
-      if (isSubType(t1, t2)) Empty
-      else if (isSubType(t2, t1) && canDecompose(t1))
-        minus(Or(getPartitions(t1).map(Typ)), b)
+    case (Typ(tp1), Typ(tp2)) =>
+      if (isSubType(tp1, tp2)) Empty
+      else if (isSubType(tp2, tp1) && canDecompose(tp1))
+        minus(Or(partitions(tp1).map(Typ)), b)
       else a
-    case (Typ(t1), Kon(t2, ss)) =>
-      if (isSubType(t1, t2)) minus(Kon(t2, getSignature(t2).map(Typ)), b)
-      else if (isSubType(t2, t1) && canDecompose(t1))
-        minus(Or(getPartitions(t1).map(Typ)), b)
+    case (Typ(tp1), Kon(tp2, ss)) =>
+      if (isSubType(tp1, tp2)) minus(Kon(tp2, signature(tp2).map(Typ)), b)
+      else if (isSubType(tp2, tp1) && canDecompose(tp1))
+        minus(Or(partitions(tp1).map(Typ)), b)
       else a
     case (_, Or(ss)) =>
       ss.foldLeft(a)(minus)
     case (Or(ss), _) =>
       Or(ss.map(minus(_, b)))
-    case (Kon(t1, ss), Typ(t2)) =>
-      if (isSubType(t1, t2)) Empty
+    case (Kon(tp1, ss), Typ(tp2)) =>
+      if (isSubType(tp1, tp2)) Empty
       else if (simplify(a) == Empty) Empty
-      else if (isSubType(t2, t1) && canDecompose(t1))
-        minus(Or(getPartitions(t1).map(Typ)), b)
+      else if (isSubType(tp2, tp1) && canDecompose(tp1))
+        minus(Or(partitions(tp1).map(Typ)), b)
       else a
-    case (Kon(t1, ss1), Kon(t2, ss2)) =>
-      if (!isEqualType(t1, t2)) a
+    case (Kon(tp1, ss1), Kon(tp2, ss2)) =>
+      if (!isEqualType(tp1, tp2)) a
       else if (ss1.zip(ss2).exists(p => simplify(intersect(p._1, p._2)) == Empty)) a
       else if (ss1.zip(ss2).forall((subspace _).tupled)) Empty
       else
         Or(
           ss1.zip(ss2).map((minus _).tupled).zip(0 to ss2.length - 1).map {
-            case (ri, i) => Kon(t1, ss1.updated(i, ri))
+            case (ri, i) => Kon(tp1, ss1.updated(i, ri))
           })
   }
 }
 
-// scala implementation of space logic
+/** Scala implementation of space logic */
 class SpaceEngine(implicit ctx: Context) extends SpaceLogic {
+  import tpd._
+
   def debug(s: String): Unit = {
     if (ctx.debug) println(s)
   }
 
-  // if roundUp, approximate extractors to its type
-  // else approximate extractors to Empty
-  def project(pat: tpd.Tree)(implicit ctx: Context, roundUp: Boolean = true): Space = pat match {
+  /** If roundUp, approximate extractors to its type, else approximate extractors to Empty */
```

Comment:
``If `roundUp` is true, approximate an extractor to its type, otherwise approximate it to `Empty` ``